autogen/pari_decl.py

```diff
@@ -0,0 +1,82 @@
+"""
+Write pari declaration files
+"""
+
+#*****************************************************************************
+#       Copyright (C) 2017 Vincent Delecroix <vincent.delecroix@labri.fr>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  http://www.gnu.org/licenses/
+#*****************************************************************************
+
+from __future__ import absolute_import
+
+import os, re
+
+from .paths import include_dirs
+
+paridecl_function_re = re.compile("[A-Za-z][^()]*[(][^()]*[)];$")
+
+paridecl_banner='''# This file is auto-generated by {}
+# from PARI header at {}
+
+cdef extern from *:
+'''
+
+def write_paridecl():
+    r"""
+    Read the file ``paridecl.h`` from the PARI distribution and write
+    ``cypari2/auto_paridecl.pxi``
+    """
+    for path in include_dirs():
+        filename = os.path.join(path, 'pari', 'paridecl.h')
+        if os.path.isfile(filename):
+            with open(os.path.join(filename)) as f:
+                text = f.read()
+                break
+    else:
+        raise ValueError('paridecl.h not found')
+
+    global paridecl_banner
+    paridecl_banner = paridecl_banner.format(__file__, filename)
+
+    outfilename = os.path.join('cypari2', 'auto_paridecl.pxi')
+    with open(outfilename, 'w') as f:
+        f.write(paridecl_banner)
+        for line in text.splitlines():
+            if line.startswith('/*') and line.endswith('*/'):
+                if not comment:
+                    f.write('\n')
+                f.write('    #')
+                f.write(line[2:-2])
+                f.write('\n')
+                comment = True
+            elif '/*' in line and '*/' in line:
+                print('WARNING: ignoring pari declaration {}'.format(line))
+            elif paridecl_function_re.match(line):
+                # struct type declarations in functions should be avoided
+                line = line.replace(' struct ', ' ')
+                line = line.replace('(struct ', '(')
+
+                # (void) should be ()
+                line = line.replace('(void);', '();')
+
+                # lambda is not a valid variable name in Cython
+                line = line.replace('long *lambda', 'long *')
```

Comment:
This is not robust! A declaration like `long *lambdaf(void);` would be wrongly modified! Would be better to check for the specific line it is intended for
```
GEN     ZX_ZXY_rnfequation(GEN A, GEN B, long *lambda);
```